File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int res=0;
        int n=nums.size();

        if(n==1)
            return nums[0];

        vector<int> dp(n,0);

        dp[0]=nums[0];
        dp[1]=nums[1];

        res=max(dp[0],dp[1]);
        dp[1]=res;
        for(int i=2;i<n;i++){
            dp[i]=max(dp[i-1], (nums[i] + dp[i-2]));

            res = max(res,dp[i]);
            cout<< dp[i]<<' '; 
        }

        return res;
    }

    // 5,1,2,10,6, 2,  7, 9,3,1
    // 5 5 7 15 15 17  22 26 26 27 
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    public int dfs(TreeNode root,int maxSoFar){
        if(root==null)
            return 0;
        int res=0;
        if(root.val >= maxSoFar)
            res++;

        maxSoFar = Math.max(maxSoFar,root.val);
        if(root.left!=null)
            res += dfs(root.left,maxSoFar);
        if(root.right!=null)
            res += dfs(root.right,maxSoFar);

        return res;    

    }
    public int goodNodes(TreeNode root) {
        return dfs(root,-105);
    }
}
